src/server/rollback_postgresql.cpp

```diff
@@ -0,0 +1,617 @@
+	prepareStatement("node_list", "SELECT id, name FROM node");
+
+	prepareStatement("actor_upsert",
+		"INSERT INTO actor(name) VALUES($1) "
+		"ON CONFLICT (name) DO UPDATE SET name = EXCLUDED.name "
+		"RETURNING id");
+	prepareStatement("node_upsert",
+		"INSERT INTO node(name) VALUES($1) "
+		"ON CONFLICT (name) DO UPDATE SET name = EXCLUDED.name "
+		"RETURNING id");
+
+	prepareStatement("action_insert",
+		"INSERT INTO action ("
+			"actor, timestamp, type,"
+			"list, index, add, stackNode, stackQuantity, nodeMeta,"
+			"x, y, z,"
+			"oldNode, oldParam1, oldParam2, oldMeta,"
+			"newNode, newParam1, newParam2, newMeta,"
+			"guessedActor"
+		") VALUES ("
+			"$1::int, $2::bigint, $3::int,"
+			"$4, $5::int, $6::int, $7::int, $8::int, $9::int,"
+			"$10::int, $11::int, $12::int,"
+			"$13::int, $14::int, $15::int, $16::bytea,"
+			"$17::int, $18::int, $19::int, $20::bytea,"
+			"$21::int"
+		")");
+
+	prepareStatement("action_select",
+		"SELECT "
+			"actor, timestamp, type,"
+			"list, index, add, stackNode, stackQuantity, nodeMeta,"
+			"x, y, z,"
+			"oldNode, oldParam1, oldParam2, oldMeta,"
+			"newNode, newParam1, newParam2, newMeta,"
+			"guessedActor "
+		"FROM action "
+		"WHERE timestamp >= $1::bigint "
+		"ORDER BY timestamp DESC, id DESC");
+
+	prepareStatement("action_select_with_actor",
+		"SELECT "
+			"actor, timestamp, type,"
+			"list, index, add, stackNode, stackQuantity, nodeMeta,"
+			"x, y, z,"
+			"oldNode, oldParam1, oldParam2, oldMeta,"
+			"newNode, newParam1, newParam2, newMeta,"
+			"guessedActor "
+		"FROM action "
+		"WHERE timestamp >= $1::bigint AND actor = $2::int "
+		"ORDER BY timestamp DESC, id DESC");
+
+	prepareStatement("action_select_range",
+		"SELECT "
+			"actor, timestamp, type,"
+			"list, index, add, stackNode, stackQuantity, nodeMeta,"
+			"x, y, z,"
+			"oldNode, oldParam1, oldParam2, oldMeta,"
+			"newNode, newParam1, newParam2, newMeta,"
+			"guessedActor "
+		"FROM action "
+		"WHERE timestamp >= $1::bigint "
+			"AND x IS NOT NULL AND y IS NOT NULL AND z IS NOT NULL "
+			"AND x BETWEEN $2::int AND $3::int "
+			"AND y BETWEEN $4::int AND $5::int "
+			"AND z BETWEEN $6::int AND $7::int "
+		"ORDER BY timestamp DESC, id DESC "
+		"LIMIT $8::int");
+}
+
+int RollbackMgrPostgreSQL::getActorId(const std::string &name)
+{
+	auto &cache = m_cache;
+	if (auto it = cache.actor_name_to_id.find(name); it != cache.actor_name_to_id.end())
+		return it->second;
+
+	verifyDatabase();
+	const char *values[] = { name.c_str() };
+
+	PGresult *r = execPrepared("actor_upsert", 1, values, false, false);
+	if (PQntuples(r) == 0) {
+		PQclear(r);
+		throw DatabaseException("PostgreSQL rollback: failed to upsert actor id");
+	}
+	const int id = pg_to_int(r, 0, 0);
+	PQclear(r);
+	cache.actor_name_to_id.emplace(name, id);
+	cache.actor_id_to_name.emplace(id, name);
+	return id;
+}
+
+int RollbackMgrPostgreSQL::getNodeId(const std::string &name)
+{
+	auto &cache = m_cache;
```

Comment:
This does neither improve readability nor performance (still 1 level of indirection). Please just use `m_cache` directly. The same comment for the other occurrences.